rma/models/rma.py

```diff
@@ -650,21 +642,125 @@ def action_rma_send(self):
+
+    def _prepare_procurement(
+        self,
+        location,
+        values,
+        product=None,
+        qty=None,
+        uom=None,
+    ):
+        self.ensure_one()
+        group = values.get("group_id")
+        product = product or self.product_id
+        if not qty:
+            qty = self.product_uom_qty
+            uom = self.product_uom
+        return self.env["procurement.group"].Procurement(
+            product,
+            qty,
+            uom,
+            location,
+            product.display_name,
+            group.name,
+            self.company_id,
+            values,
+        )
+
+    def _prepare_reception_procurement_values(self, group=None):
+        values = self._prepare_procurement_values(group=group)
+        values.update(
+            {
+                "rma_receiver_ids": [(6, 0, self.ids)],
+            }
+        )
+        if self.move_id:
+            values.update(
+                {
+                    "origin_returned_move_id": self.move_id.id,
+                }
+            )
+        return values
+
+    def _prepare_reception_procurement_group_values(self):
+        self.ensure_one()
+        return {}
+
+    def _create_reception_procurement_group(self):
+        return self._create_procurement_group(
+            self._prepare_reception_procurement_group_values()
+        )
+
+    def _prepare_reception_procurement(self):
+        self.ensure_one()
+        group = self.procurement_group_id
+        if not group:
+            group = self._create_reception_procurement_group()
+        values = self._prepare_reception_procurement_values(group)
+        return self._prepare_procurement(self.location_id, values)
+
+    def _prepare_reception_procurements(self):
+        procurements = []
+        for rma in self:
+            if not rma._product_is_storable():
+                continue
+            procurements.append(rma._prepare_reception_procurement())
+        return procurements
+
+    def _run_procurements(self, procurements):
+        if procurements:
+            self.env["procurement.group"].run(procurements)
+
+    def _create_receptions(self):
+        procurements = self._prepare_reception_procurements()
+        self._run_procurements(procurements)
 
     def action_confirm(self):
         """Invoked when 'Confirm' button in rma form view is clicked."""
-        self.ensure_one()
         self._ensure_required_fields()
-        if self.state == "draft":
-            if self.picking_id:
-                reception_move = self._create_receptions_from_picking()
-            else:
-                reception_move = self._create_receptions_from_product()
-            self.write({"reception_move_id": reception_move.id, "state": "confirmed"})
-            self._add_message_subscribe_partner()
-            self._send_confirmation_email()
+        self = self.filtered(lambda rma: rma.state == "draft")
+        if not self:
+            return
+        self._create_receptions()
```

Comment:
I don't think we should extract this method too. When a real example requires it, then it can be extracted. Meanwhile, the code gets very complicated to be followed with all of this and with no real gain.

Reply:
This method was already extracted previously (_create_receptions_from_picking and _create_receptions_from_product).